Add -fdiagnostics-show-caret to display the context of an error

lefessan opened the pull request:
When a warning or error is printed, prints the source context with the line of the error, plus 2 lines before and after.
COBC_PREVIEW_ERROR can be used to set the option globally.

lefessan commented on tests/testsuite.src/used_binaries.at:
```diff
@@ -793,19 +793,40 @@ AT_CLEANUP
+])
+
+# check -fpreview-error prints error context
+AT_CHECK([$COMPILE -j -fpreview-error prog.cob], [1], [], [prog.cob:6: error: syntax error, unexpected Identifier
```
Here is an example...

lefessan commented:
As an example:
```
prompt> $COMPILE -j -fpreview-error prog.cob
prog.cob:6: error: syntax error, unexpected Identifier

  0004          PROCEDURE        DIVISION.
  0005              DISPLAY "jobby"
  0006 >            EXIT HERE.

```

lefessan commented:
Note that the default is currently to disable this feature, but it could be activated by default (and disabled through `COBC_PREVIEW_ERROR=0` for the testsuites)

GitMensch commented:
Hm, the idea is good and the general point of reworking the diagnostic format is even on the GSOC idea list for GnuCOBOL (did you get it from there/the note on the dev list about that?).

But reading the source file again should be absolute last resort and will be very likely wrong for REPLACE/REPLACING and won't work for complies from stdin.

So... Could you please rework that to possibly save the context from the scanner via to a static buffer?

Also note that -fno-diagnostics-show-line-numbers should be added along with this change (which would drop the leading line numbers) and possibly -fdiagnostics-show-caret should be used to toggle this instead of -fpreview-error ... and ideally we should also use the caret to show where the error is.

As a general rule of thumb we'd like to reach https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Message-Formatting-Options.html ...
someday.

lefessan commented:
> Hm, the idea is good and the general point of reworking the diagnostic format is even on the GSOC idea list for GnuCOBOL (did you get it from there/the note on the dev list about that?). 

No, I didn't know about the GSOC idea. Is it discussed on the forum ?
I did that patch because I needed it myself... which is usually how I work, as I can only devote a small amount of time on the project from time to time.

> But reading the source file again should be absolute last resort and will be very likely wrong for REPLACE/REPLACING and won't work for complies from stdin. So... Could you please rework that to possibly save the context from the scanner via to a static buffer? 
> Also note that -fno-diagnostics-show-line-numbers should be added along with this change (which would drop the leading line numbers) and possibly -fdiagnostics-show-caret should be used to toggle this instead of -fpreview-error ... and ideally we should also use the caret to show where the error is. 

I won't be able to spend much more time on this patch. For now, I changed the name of the option to `-fdiagnostics-show-caret` and added another option to disable it for warnings (typically, we are working on a project where there are currently a lot of warnings, and we want to only focus on errors in a first step).

> As a general rule of thumb we'd like to reach https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Message-Formatting-Options.html ... someday.

Would it be ok to merge this work as it is, and wait for the GSOC to get it improved ? I will be working on adding a bit more GCOS support now, so I won't be able to improve this patch more before some time.

lefessan commented:
@GitMensch Hi, I say on the forum that you mentioned this patch. Do you think it would be ok to merge it in the current state ?

GitMensch commented on cobc/error.c:
```diff
@@ -63,10 +70,73 @@ print_error_prefix (const char *file, int line, const char *prefix)
+	     && file != NULL && line
+	     && ( cb_diagnostics_warning_show_caret || kind == CB_KIND_ERROR )
+		){
+		FILE* fd = fopen(file, "r");
```
needs to check `fd` after open and likely should check before and not do that if name matches `COB_DASH`

lefessan replied:
I added checks for NULL and COB_DASH.

GitMensch reviewed (changes requested):
> @GitMensch Hi, I've seen on the forum that you mentioned this patch. Do you think it would be ok to merge it in the current state ?

Not in its current state :-/ But after reviewing the [initial GCC implementation](https://github.com/gcc-mirror/gcc/commit/9fec00429dd36ea51798afeaf5f6b9c1987b85cd) (also opened the file then read and print the lines), I'm now open to add it without the design change I previously thought of (GCC did that later on).

Things that are likely easy to do:

* drop `COBC_DIAGNOSTICS_SHOW_CARET` completely; for the testsuite it is enough to add `-fno-diagnostics-show-caret` in tests/atlocal.in, just as I did with `-fno-diagnostics-show-option`
* rebase
* add `-fno-diagnostics-show-line-numbers` (because it is just a single flag that only modifies the output in one place)
* add an entry in NEWS and doc/gnucobol.tex about the two new options

Then, as a follow up, we may [use source caches, like the current GCC implementation does](https://github.com/gcc-mirror/gcc/blob/5d85b5d649fff675ff00adcc99371bccf4ef5944/gcc/input.cc#L858).

But still you may want to have a look at the old diagnostic print function linked before your commit and look for good style; and have a look at [the current one](https://github.com/gcc-mirror/gcc/blob/5d85b5d649fff675ff00adcc99371bccf4ef5944/gcc/diagnostic-show-locus.cc#L2831)

lefessan commented on cobc/cobc.c:
```diff
@@ -8648,6 +8648,18 @@ set_cobc_defaults (void)
 		 *p == '1')) {
 		cb_unix_lf = 1;
 	}
+	p = cobc_getenv ("COBC_DIAGNOSTICS_SHOW_CARET");
```
I want to argue to keep it: most of the options in the compiler do not need an env variable because they are project-dependent, i.e. the project developers should whether these options should be set or not for their project. On the contrary, the way errors are displayed is user-dependent, not project-dependent. As such, users may want to set or unset these options in their config files, so that they are used by `make` automatically. So the env variable is needed for that (unless we add a user config file for gnucobol).

GitMensch replied:
I see the point - but we shouldn't do this for a single option, a user may want to have the diagnostic name or not, he maybe even don't care about any warnings, ...
A project specific setup that handles this would set default flags in the makefile and passes additional flags from the user's environment or include file (depending on the setup either before or after the default makefile flags).

Actually the way projects can nicely be setup via Makefiles is the main reason for not having a global env var or a config file (other than dialect options). There's a FR about this on SF if you want to give that more thoughts - for this PR, please drop that.

lefessan commented on cobc/flag.def:
```diff
@@ -232,6 +232,16 @@ CB_FLAG (cb_listing_symbols, 1, "tsymbols",
+	  "                               (the source is displayed without any REPLACE performed."
+	))
+
+CB_FLAG_ON (cb_diagnostics_warning_show_caret, 1, "diagnostics-show-caret-on-warning",
```
I removed it.

GitMensch commented on NEWS:
```diff
@@ -325,12 +325,19 @@ NEWS - user visible changes				-*- outline -*-
-   were not usable, for example SYNCHRONIZED or COMPUTATIONAL. This was fixed
-   and reserved words updated for the dialects "acu" (to ACUCOBOL-GT 10.4),
-   "ibm" (to Enterprise COBOL 6.3) and "mf" (to Micro Focus Visual COBOL 6.0)
+   were not usable, for example SYNCHRONIZED or COMPUTATIONAL. This
```
unrelated change?

GitMensch commented on cobc/error.c:
```diff
@@ -63,10 +70,80 @@ print_error_prefix (const char *file, int line, const char *prefix)
+				printed = 1;
+				fprintf (stderr, "  ");
+				if (cb_diagnostics_show_line_numbers){
+					fprintf (stderr, "%04d ",
```
Currently that does not only skip line number but the whole printing, no?

lefessan replied:
Wrong indentation, sorry...

GitMensch commented on tests/testsuite.src/used_binaries.at:
```diff
@@ -918,3 +912,100 @@ AT_CHECK([cat prog.cob | $COMPILE_MODULE -j -], [0], [job], [],
+
+prog.cob:6: warning: numeric value is expected
+])
+AT_CHECK([$COMPILE -fdiagnostics-show-caret -fno-diagnostics-show-caret-on-warning -j prog.cob], [1], [], [prog.cob:7: error: CRUD.CPY: No such file or directory
```
`-fno-diagnostics-show-caret-on-warning` doesn't exists any more, please replace it with a test for `-fno-diagnostics-show-line-numbers`

lefessan replied:
done, replaced with a test on `-fno-diagnositcs-show-line-numbers`

GitMensch reviewed (changes requested):
Nearly ready now - see review details.

GitMensch commented:
freindly ping @lefessan

lefessan commented:
I think all your requests have been fixed.

GitMensch commented on cobc/flag.def:
```diff
@@ -232,6 +232,16 @@ CB_FLAG (cb_listing_symbols, 1, "tsymbols",
+	))
+
+CB_FLAG_ON (cb_diagnostics_show_line_numbers, 1, "diagnostics-show-line-numbers",
+	_("  -fno-diagnostics-show-line-numbers     suppress display of line numbers in diagnostics"
```
"strange" thing to do, but please start the second line on the same position that is used for `diagnostics-show`... and as the name is long delimit the first line with \t instead of spaces

GitMensch commented on tests/testsuite.src/used_binaries.at:
```diff
@@ -794,23 +794,17 @@ AT_CLEANUP
-
-
-AT_DATA([prog.cob], [
+AT_DATA([prog.cob],[
```
please drop these and the following unrelated changes (enough during commit)

GitMensch commented on tests/testsuite.src/used_binaries.at:
```diff
@@ -918,3 +912,62 @@ AT_CHECK([cat prog.cob | $COMPILE_MODULE -j -], [0], [job], [],
+           END-DISPLAY
+           STOP RUN.
+])
+AT_CHECK([$COBC -Wall -fsyntax-only prog.cob], [1], [], [prog.cob:7: error: CRUD.CPY: No such file or directory
```
this line should be:

```
AT_CHECK([$COMPILE_ONLY prog.cob], [1], [],
[prog.cob:7: error: CRUD.CPY: No such file or directory
```

the following test should be

```
AT_CHECK([$COMPILE_ONLY -fdiagnostics-show-caret -fdiagnostics-show-line-numbers prog.cob], [1], [],
[prog.cob:7: error: CRUD.CPY: No such file or directory
```

and the last (which may be switched with the second)

```
AT_CHECK([$COMPILE_ONLY -fdiagnostics-show-caret -prog.cob], [1], [],
[prog.cob:7: error: CRUD.CPY: No such file or directory
```

GitMensch commented on cobc/flag.def:
```diff
@@ -232,6 +232,16 @@ CB_FLAG (cb_listing_symbols, 1, "tsymbols",
 	_("  -fno-diagnostics-show-option\tsuppress output of option that directly\n"
 	  "                        controls the diagnostic"))
+
+CB_FLAG (cb_diagnostics_show_caret, 1, "diagnostics-show-caret",
```
If possible please change to `CB_FLAG_ON` and add the reverse to the flags in tests/atlocal.in

GitMensch reviewed (changes requested):
only minor changes needed (feel free to directly commit)

If you want to finish that completely, then add `-fdiagnostics-plain-output` to cobc.c (this is a "normal" option, no flag, it has no "no" and would set both new options to ´0`) and use this in tests/atlocal.in for the flags.

GitMensch commented:
Possibly use in the NEWS something similar to GCCs entry:

the diagnostics now print the source code context with a left margin showing line numbers, configurable with -fno-diagnostics-show-line-numbers, and possible to disable completely with -fno-diagnostics-show-caret;

the option -fdiagnostics-plain-output was added to request that diagnostic output look as plain as possible and stay more stable over time

lefessan commented on cobc/error.c:
```diff
@@ -63,10 +70,80 @@ print_error_prefix (const char *file, int line, const char *prefix)
+/* Display a context around the location of the error/warning, only if
+ * cb_diagnostics_show_caret is true
+
+   Only display two lines before and after. No caret yet for the column as
```
My intuition is that it's better to display the file as it is, so that it is easy for the developer to visually match the error location in a terminal with the one in an editor.

GitMensch replied:
Totally reasonable

lefessan replied:
My intuition is that it's better to display the file as it is, so that it is easy for the developer to visually match the error location in a terminal with the one in an editor.

GitMensch reviewed (commented):
the last commit pushed LGTM

lefessan commented:
I think I added all your suggestions, I will commit it to svn as soon as CI is passed. Thanks !

lefessan commented:
Merged in SVN

GitMensch commented:
Sadly that broke C89 support in cobc/error.c: https://ci.appveyor.com/api/buildjobs/vqa49l488as7peee/log

Please adjust to declare the variables at the beginning of the scope.

Apart from that: what do you think of adding a std=c89 to the CI here? That _should_ have likely called that.

lefessan commented:
Sorry, I didn't see your comment, I am going to fix this problem asap

GitMensch commented:
Please drop a note if you think you'll have moved around the declarations in the next two hours as I'm doing it otherwise afterwards.

GitMensch commented:
Fixing C89 and additional changes to the format and some of the details done "upstream" in [svn r5081](https://sourceforge.net/p/gnucobol/code/5081/).

CI PR already referenced.

GitMensch commented:
Can you have a look for some late-cleanup on this, moving ` -fdiagnostics-plain-output` from `COBC` to `COBOL_FLAGS`, please?
I'd like to keep the command line for NIST, which uses `COBC`, as small as possible.

GitMensch commented:
In the meantime I've did that adjustment upstream.
